Issue #66: Add search API

TomoTsuyuki opened the pull request:
Implement search area for customfield, which can be set in the "Search setting".

TomoTsuyuki committed:
Issue #66: Add search API

dmitriim commented on settings.php:
```diff
@@ -56,5 +56,23 @@
+        get_string('search:settings:area', 'mod_cms'),
+        get_string('search:settings:area_desc', 'mod_cms'),
+        null,
+        $areas,
```
php 7.1 doesn't support this extra comma in the end

TomoTsuyuki replied:
The comma has been removed.

TomoTsuyuki replied:
Actually, the setting has been removed.

dmitriim commented on classes/search/cms.php:
```diff
@@ -0,0 +1,188 @@
+ * @copyright  2024 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class cms extends \core_search\base_mod {
```
any reason for picking cms class  name? Looks like other activities have it as 'activity' e.g. mod_assign\search\activity

TomoTsuyuki replied:
I checked other activity modules, and they are using record name - e.g. `chapter`, `forum`, etc.
I renamed it to `cmsfield` as it's from cmsfield custom field handler.

dmitriim commented on classes/search/cms.php:
```diff
@@ -0,0 +1,188 @@
+     */
+    public function get_context_url(\core_search\document $doc) {
+        $contextmodule = \context::instance_by_id($doc->get('contextid'));
+        return new \moodle_url('/mod/cms/view.php', ['id' => $contextmodule->instanceid]);
```
I would suggest to double what the client expects to see as a link to found documents in the search results.

TomoTsuyuki replied:
This is just link and actual result is in doc https://github.com/catalyst/moodle-mod_cms/pull/129/files#diff-965e63736ed197b659eeb7349499abdfc8e2aa372cf1fd0a8ddb6b3467f1bf5aR98

dmitriim replied:
Yes and what about the link in the search result? It should link to some page, shouldn't it?

TomoTsuyuki replied:
I'm asking to the client what they expect.
Meanwhile, I'll leave same here and update other parts of this PR.

dmitriim commented on settings.php:
```diff
@@ -56,5 +56,23 @@
+    $areas = $DB->get_records_sql_menu($sql);
+
+    $settings->add(new admin_setting_configmulticheckbox(
+        'mod_cms/search_area',
```
why did you pick to implement search areas like that? How would it work in the search results whey you disable one of the areas here, but documents from this area are already indexed?

TomoTsuyuki replied:
It can be complicated logic, and I removed this setting.

TomoTsuyuki committed:
Issue #66: Rename search class and debug for php7.2

TomoTsuyuki committed:
Issue #66: Remove setting and add phpunit test

TomoTsuyuki committed:
Issue #66: Add phpunit test for check_access

dmitriim commented on lang/en/cms.php:
```diff
@@ -103,6 +103,12 @@
 
+// Search strings.
+$string['search:cmsfield'] = 'CMS';
+$string['search:settings'] = 'Search settings';
```
Looks like those strings are not used anymore.

TomoTsuyuki replied:
It was used for search area settings, but the settings have been removed.
I just removed this lang string.

dmitriim commented on classes/search/cmsfield.php:
```diff
@@ -0,0 +1,187 @@
+                     WHERE mcd.id = :id";
+            $this->cmsdata[$id] = $DB->get_record_sql($sql, ['id' => $id]);
+            if (!$this->cmsdata[$id]) {
+                throw new \dml_missing_record_exception('cms_data');
```
Can you please explain why we are throwing dml_missing_record_exception exception here?

TomoTsuyuki replied:
I updated the `$DB->get_record_sql` to use `MUST_EXIST` option which is using `dml_missing_record_exception` in it.

dmitriim commented on lang/en/cms.php:
```diff
@@ -103,6 +103,12 @@
 $string['error:no_config_hash'] = 'Module {$a} has no config hash.';
 
+// Search strings.
+$string['search:cmsfield'] = 'CMS';
```
Can you please bump version?

TomoTsuyuki replied:
Version bumped.

dmitriim commented on classes/search/cmsfield.php:
```diff
@@ -0,0 +1,187 @@
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
```
this doc block is not required anymore.

TomoTsuyuki replied:
Removed.

dmitriim commented on classes/search/cmsfield.php:
```diff
@@ -0,0 +1,187 @@
+     *
+     * @param int $modifiedfrom timestamp
+     * @param \context|null $context Optional context to restrict scope of returned results
+     * @return moodle_recordset|null Recordset (or null if no results)
```
should be \moodle_recordset

TomoTsuyuki replied:
Updated.

dmitriim commented on classes/search/cmsfield.php:
```diff
@@ -0,0 +1,187 @@
+            return \core_search\manager::ACCESS_DENIED;
+        }
+
+        $context = \context_module::instance($cminfo->id);
```
Should we move this under line 122 same way it's done on line 83? So we get ACCESS_DELETED in case context is missing for some reason?

TomoTsuyuki replied:
moved into the `try` block

dmitriim commented on classes/search/cmsfield.php:
```diff
@@ -0,0 +1,187 @@
+
+        $context = \context_module::instance($cminfo->id);
+
+        if (!has_capability('mod/cms:view', $context)) {
```
What about 'mod/cms:seeall' capability? Do we need to check it as well?

TomoTsuyuki replied:
`mod/cms:seeall` is for showing CMS types in the activity chooser.
I don't think we need to use for search.

dmitriim commented on tests/search/search_test.php:
```diff
@@ -0,0 +1,234 @@
+
+        $this->getDataGenerator()->enrol_user($user1->id, $course->id, 'student');
+
+        // Name for cms is coming from "title_mustache" in cms_type.
```
What does this comment mean? How is it relevant to the code?

TomoTsuyuki replied:
I updated the comment.
Basically, any activity should have "name".
But for cms activity, the name is coming from "title_mustache" in cms_type table.
I was confused about the name, so I added the comment.

dmitriim commented on tests/search/search_test.php:
```diff
@@ -0,0 +1,234 @@
+        $this->assertEquals(2, iterator_count($recordset));
+        $recordset->close();
+
+        // The +2 is to prevent race conditions.
```
Can you please explain a bit more about this comment? What race conditions we are trying to avoid?

TomoTsuyuki replied:
It's just copied from other plugin actually.
I updated the comment.
It's just change the time to future to make sure it's empty as there is no record after that time.

TomoTsuyuki committed:
Issue #66: Version bump

dmitriim commented on classes/search/cmsfield.php:
```diff
@@ -0,0 +1,187 @@
+                  JOIN {customfield_field} mcf ON mcf.id = mcd.fieldid
+                  JOIN {customfield_category} mcc ON mcf.categoryid = mcc.id
+          $contextjoin
+                 WHERE mcd.timemodified >= ? AND mcc.component = 'mod_cms' AND mcc.area = 'cmsfield'
```
Do we also need to filter by supported field types?

TomoTsuyuki replied:
Type added.

TomoTsuyuki committed:
Issue #66: Add default value and update unit test

TomoTsuyuki committed:
Issue #66: Add comments and version bump

TomoTsuyuki commented:
I rebased from the latest code, and also added indexing from default values.
The search result can be updated from requirement, but the code can be reviewed other things for now.

dmitriim commented on classes/search/cmsfield.php:
```diff
@@ -0,0 +1,231 @@
+        }
+
+        // Get default value for cms custom field.
+        if (is_null($this->defaultvalues)) {
```
If you abstract this part to get_default_values method it would be much cleaner.

TomoTsuyuki replied:
This part has been moved to the function `get_default_values`

TomoTsuyuki committed:
Issue #66: Fix CI issues

dmitriim commented on classes/search/cmsfield.php:
```diff
@@ -0,0 +1,231 @@
+          $contextjoin
+                 WHERE mcd.timemodified >= ? AND mcc.component = 'mod_cms' AND mcc.area = 'cmsfield'
+                   AND mcf.type IN ('textarea', 'text')
+                 UNION
```
with using union, wouldn't we get [mc.id](http://mc.id/) duplicated in the first column?

TomoTsuyuki replied:
Yes, it is possible.
I updated the SQL to concatenate contents, so both contents can be searched.

TomoTsuyuki committed:
Issue #66: Debug search result

TomoTsuyuki committed:
Issue #66: Concatenate texts in same activity

TomoTsuyuki committed:
Issue #66: Set support version to 4.1

TomoTsuyuki commented:
The groupconcat function in Moodle is available from 3.11.
Currently, Moodle supports from 4.1, so I dropped this plugin support under 4.0.
The bug in the search has been fixed.
Could you review and test again @dmitriim ?

TomoTsuyuki committed:
Issue #66: Update README.md
